File: frontend_flutter/lib/src/models/user_model.dart
import 'dart:convert';

class User {
  final String id;
  final String name;
  final String email;
  final String phone;
  final String? address;
  final String role;
  final String? image;

  User(
      {required this.id,
      required this.name,
      required this.phone,
      required this.email,
      required this.role,
      this.image,
      this.address});

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'email': email,
      'phone': phone,
      'role': role,
      'image': image,
      'address': address,
    };
  }

  factory User.fromMap(Map<String, dynamic> map) {
    return User(
      id: map['id'],
      name: map['name'],
      email: map['email'],
      phone: map['phone'],
      role: map['role'],
      image: map['image'],
      address: map['address'],
    );
  }

  String toJson() => json.encode(toMap());

  factory User.fromJson(source) => User.fromMap(source);
}

File: server_nodejs/src/models/user.model.js
import mongoose from 'mongoose';


//create User schema
const UserSchema = new mongoose.Schema({
    name: {
        type: String,
        required: true
    },
    email: {
        type: String,
        index: true,
        required: true
    },
    password: {
        type: String,
        required: true
    },
    gender: {
        type: String,
        enum: ['male', 'female'],
        required: false
    },
    phone: {
        type: String,
        required: false
    },
    image: {
        type: String,
        required: false
    },

    address: {
        type: String,
        required: false
    },
    created_at: {
        type: Date,
        default: Date.now
    },
    updated_at: {
        type: Date,
        default: Date.now
    },
    role: {
        default: 'customer',
        type: String,
        enum: ['admin', 'customer', 'shopOwner'],
        required: true
    }


});

UserSchema.set("toJSON", {
    transform: (document, returnedObject) => {
        returnedObject.id = returnedObject._id.toString();
        delete returnedObject._id;
        delete returnedObject.__v;
        //do not reveal passwordHash
        delete returnedObject.password;
    },
});

const User = mongoose.model('User', UserSchema);

// User.create({
//     name: 'John',
//     email: 'john@example.com',
//     password: 'password',
//     gender: 'male'
// });

module.exports = User;